Adding the json script tests from ABC

Danconnolly opened the pull request:
This pull request incorporates the standard script unit tests from the script_tests.json file. These unit tests are used by the major node implementations. Care has been taken to avoid modifying the script_tests.json file (with the exception of attribution), so that it can be re-imported in the future as new tests are added.

The tests have shown several deficiencies in the BitcoinJ Script implementation. I have corrected enough to get the basic tests working but many are still not working. Tests which cover features that have not been implemented in bitcoinj-cash yet are ignored. See the TODO notes in  ScriptDataDrivenTest.java.

These broken tests need to be addressed at some point. I have run out of time to devote on this right now and am posting it here as a "work in progress".

Danconnolly committed:
define several missing constants - CHECKSEQUENCEVERY, MINIMALIF, NULLFAIL, SIGHASH_FORKID, REPLAY_PROTECTION

Danconnolly committed:
define several missing constants - CHECKSEQUENCEVERY, MINIMALIF, NULLFAIL, SIGHASH_FORKID, REPLAY_PROTECTION

Danconnolly committed:
define several missing constants - CHECKSEQUENCEVERY, MINIMALIF, NULLFAIL, SIGHASH_FORKID, REPLAY_PROTECTION

Danconnolly committed:
add the script_tests.json file from ABC and use for script tests

Danconnolly committed:
add the script_tests.json file from ABC and use for script tests

Danconnolly committed:
add the script_tests.json file from ABC and use for script tests

Danconnolly committed:
add ScriptException sub-classes for specific errors

Danconnolly committed:
add ScriptException sub-classes for specific errors

Danconnolly committed:
add ScriptException sub-classes for specific errors

Danconnolly committed:
rename VerifyException to OpVerifyFailed, to avoid confusion with VerificationException

Danconnolly committed:
rename VerifyException to OpVerifyFailed, to avoid confusion with VerificationException

Danconnolly committed:
rename VerifyException to OpVerifyFailed, to avoid confusion with VerificationException

Danconnolly committed:
sort the definition of the exceptions

Danconnolly committed:
sort the definition of the exceptions

Danconnolly committed:
sort the definition of the exceptions

Danconnolly committed:
add SignatureFormatError exception

Danconnolly committed:
add SignatureFormatError exception

Danconnolly committed:
add SignatureFormatError exception

Danconnolly committed:
dont swallow exceptions

Danconnolly committed:
dont swallow exceptions

Danconnolly committed:
dont swallow exceptions

Danconnolly committed:
update json script tests from ABC

Danconnolly committed:
update json script tests from ABC

Danconnolly committed:
update json script tests from ABC

Danconnolly committed:
add tests of encoded signature before trying to decode it, and catch empty signature

Danconnolly committed:
add tests of encoded signature before trying to decode it, and catch empty signature

Danconnolly committed:
add tests of encoded signature before trying to decode it, and catch empty signature

Danconnolly committed:
update copyright information

Danconnolly committed:
update copyright information

Danconnolly committed:
update copyright information

Danconnolly commented:
rebased off cash-0.14 branch

Danconnolly committed:
imported new version of script_tests.json from ABC, contains improved NUM2BIN tests

Danconnolly committed:
imported new version of script_tests.json from ABC, contains improved NUM2BIN tests

Danconnolly committed:
imported new version of script_tests.json from ABC, contains improved NUM2BIN tests

Danconnolly committed:
Adjust exceptions to match new tests

The new NUM2BIN tests in script_tests.json were expecting an IMPOSSIBLE_ENCODING error message.
Added new exception EncodingMessage and adjusted NUM2BIN interpreter to produce this exception.
Exceptions were previously being generated but were the wrong type.

Danconnolly committed:
Adjust exceptions to match new tests

The new NUM2BIN tests in script_tests.json were expecting an IMPOSSIBLE_ENCODING error message.
Added new exception EncodingMessage and adjusted NUM2BIN interpreter to produce this exception.
Exceptions were previously being generated but were the wrong type.

Danconnolly committed:
Adjust exceptions to match new tests

The new NUM2BIN tests in script_tests.json were expecting an IMPOSSIBLE_ENCODING error message.
Added new exception EncodingMessage and adjusted NUM2BIN interpreter to produce this exception.
Exceptions were previously being generated but were the wrong type.

ilozanof committed:
Merge branch 'jsonscripttests' of https://github.com/nchain-research/bitcoinj-cash into jsonscripttests

Commit:
testing op return support

Commit:
testing op return support

ilozanof committed:
Implementation of MINIMALDATA, MINIMALIF, NULLFAIL and CHECKSEQUENCEVERIFY. Based on Bitcoin-abc implementation.

ilozanof committed:
Implementation of MINIMALDATA, MINIMALIF, NULLFAIL and CHECKSEQUENCEVERIFY. Based on Bitcoin-abc implementation.

ilozanof committed:
Implementation of MINIMALDATA, MINIMALIF, NULLFAIL and CHECKSEQUENCEVERIFY. Based on Bitcoin-abc implementation.

ilozanof committed:
fixing typo

ilozanof committed:
fixing typo

ilozanof committed:
fixing typo

ilozanof committed:
Updated, to cover all the tests and not "mask" the ones failing. Only one scenrio is still not covered

ilozanof committed:
Updated, to cover all the tests and not "mask" the ones failing. Only one scenrio is still not covered

ilozanof committed:
Updated, to cover all the tests and not "mask" the ones failing. Only one scenrio is still not covered

ilozanof committed:
Some code in the try-catch commented (temporary), until rest of the code is uploaded.

ilozanof committed:
Some code in the try-catch commented (temporary), until rest of the code is uploaded.

ilozanof committed:
Some code in the try-catch commented (temporary), until rest of the code is uploaded.

ilozanof committed:
Refactoring. Removing all the ScriptException subclasses. Instead, we are now using the "ScriptError" where all the possible causes for the Script failure can be stored and organized. This way, the Script failure always throws an ScriptException (easier to managed). This is how it's been implementrd in bitcoinJ (reference).

ilozanof committed:
Refactoring. Removing all the ScriptException subclasses. Instead, we are now using the "ScriptError" where all the possible causes for the Script failure can be stored and organized. This way, the Script failure always throws an ScriptException (easier to managed). This is how it's been implementrd in bitcoinJ (reference).

ilozanof committed:
Refactoring. Removing all the ScriptException subclasses. Instead, we are now using the "ScriptError" where all the possible causes for the Script failure can be stored and organized. This way, the Script failure always throws an ScriptException (easier to managed). This is how it's been implementrd in bitcoinJ (reference).

ilozanof committed:
The Hashtype Verification is taken apart in a separate function, since there is an specific error that needs to be captured and processed separately.

ilozanof committed:
The Hashtype Verification is taken apart in a separate function, since there is an specific error that needs to be captured and processed separately.

ilozanof committed:
The Hashtype Verification is taken apart in a separate function, since there is an specific error that needs to be captured and processed separately.

ilozanof committed:
Some flags related to BIP66 (relative Locks) have been added.

ilozanof committed:
Some flags related to BIP66 (relative Locks) have been added.

ilozanof committed:
Some flags related to BIP66 (relative Locks) have been added.

ilozanof committed:
Test cases fix.
These scenarios have been fixed (Expected results from script_tests.json): SIG_DER, ILLEGAL_FORKID, SIG_HASHTYPE, SIG_PUSHONLY, CLEANSTACK, PUBKEYTYPE.

ilozanof committed:
Test cases fix.
These scenarios have been fixed (Expected results from script_tests.json): SIG_DER, ILLEGAL_FORKID, SIG_HASHTYPE, SIG_PUSHONLY, CLEANSTACK, PUBKEYTYPE.

ilozanof committed:
Test cases fix.
These scenarios have been fixed (Expected results from script_tests.json): SIG_DER, ILLEGAL_FORKID, SIG_HASHTYPE, SIG_PUSHONLY, CLEANSTACK, PUBKEYTYPE.

ilozanof committed:
All the reference to Segregated Witness errors removed.

ilozanof committed:
All the reference to Segregated Witness errors removed.

ilozanof committed:
All the reference to Segregated Witness errors removed.

ilozanof committed:
added comments for clarity

ilozanof committed:
added comments for clarity

ilozanof committed:
added comments for clarity

ilozanof committed:
fixed a bug in the "isPushOnly" method, whish failes when the script has no ScriptChunks in it.

ilozanof committed:
fixed a bug in the "isPushOnly" method, whish failes when the script has no ScriptChunks in it.

ilozanof committed:
fixed a bug in the "isPushOnly" method, whish failes when the script has no ScriptChunks in it.

ilozanof committed:
changes to make it compatible with other JSON files from other repositories (like bitcoinj or bitcopinj-cash), since the structure of the tests might be a bit different in those cases.

ilozanof committed:
changes to make it compatible with other JSON files from other repositories (like bitcoinj or bitcopinj-cash), since the structure of the tests might be a bit different in those cases.

ilozanof committed:
changes to make it compatible with other JSON files from other repositories (like bitcoinj or bitcopinj-cash), since the structure of the tests might be a bit different in those cases.

ilozanof committed:
small change (1.6 jdk compatibility)

ilozanof committed:
small change (1.6 jdk compatibility)

ilozanof committed:
small change (1.6 jdk compatibility)

ilozanof committed:
added the hastype verification into the more general encoding verification

ilozanof committed:
added the hastype verification into the more general encoding verification

ilozanof committed:
added the hastype verification into the more general encoding verification

ilozanof committed:
throwing a ScriptException when a disabled opcode found, instead of an IllegalOperation

ilozanof committed:
throwing a ScriptException when a disabled opcode found, instead of an IllegalOperation

ilozanof committed:
throwing a ScriptException when a disabled opcode found, instead of an IllegalOperation

ilozanof committed:
fixed a bug in the CLEANSTACK verification flag.

ilozanof committed:
fixed a bug in the CLEANSTACK verification flag.

ilozanof committed:
fixed a bug in the CLEANSTACK verification flag.

ilozanof committed:
small change

ilozanof committed:
small change

ilozanof committed:
small change

ilozanof committed:
Merge branch 'jsonscripttests' of https://github.com/nchain-research/bitcoinj-cash into jsonscripttests

HashEngineering commented:
one of the commits #f5c8eca is for Java 1.6 compatibility, but we had another pull request that was approved for Java 1.7.

Danconnolly commented:
Good point, thanks.

Danconnolly committed:
Revert "small change (1.6 jdk compatibility)"

This reverts commit f5c8ecacc3ef4ab7faf3113c9be6bde893d565c1.

Danconnolly committed:
fix tests

Danconnolly committed:
correct spending tx in data driven tests

ilozanof committed:
Merge branch 'jsonscripttests' of https://github.com/nchain-research/bitcoinj-cash into jsonscripttests

# Conflicts:
#	core/src/main/java/org/bitcoinj/script/ScriptError.java
#	core/src/test/java/org/bitcoinj/script/ScriptHelpers.java

ilozanof committed:
Workaround for passing some test case scenarios, until the test data is changed and new data is generated (and compatible with the new Verifications implemented in the Script Engine).

ilozanof committed:
Refinement of the Workaround put in place to keep the performance of legacy tests.

Until all the legacy tests are refactored, we change slightly the ALL_VERIFY_FLAGS Collection to remove those Flags that are not compatible(SIGHASH_FORKID and PUBKEYTYPE).

ilozanof committed:
Fixed a problem in the verification of the Public Key compression verification

ilozanof committed:
Changed the comments regarding the ALL_VERIFY_FLAGS.

After looking into the code, I think the workaround (removing some Flags from ALL_VERIFY_FLAGS) is not a workaround, but a valid and needed change, in order to keep the existing functionality.

Still, I consider that a deep research into the code is needed, in order to refactor all the uses of these Flags. Since these Flags affect the way the script is executed, the different invocations to the script engine should be aware of what flags are being used 8as a parameter, for example). At this moment, these Flags are hardcoded, and that shouldn,t be the case. If we want to harcode the Verification Flags used when validating the script, then we shoul
Still, I consider that a deep research into the code is needed, in order to refactor all the uses of these Flags. Since these Flags affect the way the script is executed, the different invocations to the script engine should be aware of what flags are being used 8as a parameter, for example). At this moment, these Flags are hardcoded in some places, and that shouldn't be the case (unles there is a clear consensus about that).

The current commit runs OK and the legacy tests also run OK. But we should make it clear that ALL_VERIFY_FLAGS doen NOT measn ALL of them, just the ones that allow a "safe" script execution.

ilozanof committed:
Changed the comments in the last commit (no relevant changes in the code).

After looking into the code, I think the workaround (removing some Flags from ALL_VERIFY_FLAGS) is not a workaround, but a valid and needed change, in order to keep the existing functionality.

Still, I consider that a deep research into the code is needed, in order to refactor all the uses of these Flags. Since these Flags affect the way the script is executed, the different invocations to the script engine should be aware of what flags are being used (as a parameter, for example). At this moment, these Flags are hardcoded, and that should not be the case (unless there is a clear consensus about that).

The current commit runs OK and the legacy tests also run OK. But we should make it clear that ALL_VERIFY_FLAGS doen NOT measn ALL of them, just the ones that allow a "safe" script execution.

Commit:
testing raw import and export

Commit:
Merge remote-tracking branch 'nchain/cash-0.14' into nchain-cash-0.14

ilozanof committed:
Merge branches 'cash-0.14' and 'jsonscripttests' of https://github.com/nchain-research/bitcoinj-cash into jsonscripttests

ilozanof committed:
Added OP_MUL (re-enabled) and Tests

November 15, 2018 Hard fork. Code from SV branch used as a reference.

ilozanof committed:
Added OP_INVERT(re-enabled) and Tests

November 15, 2018 Hard fork. Code from SV branch used as a reference.
the test classes have also been refactored, to make them more flexible
for future OpCodes test coverage.

ilozanof committed:
Added more test coverage for OP_MUL and OP_INVERT

A new class "StackItem" has been added to add more flexibility for
testing, and the test classes have been refactorized.

ilozanof committed:
Added more  functionality to ScriptItem

It makes it easier to perform tests on OP_CODEs. It includes more "casting" mehods, to allow to use it with different data types or fortmats /binary, hex, numbers, etc).

ilozanof committed:
Added LSHIFT, RSHIFT and tests and refactoring

Includes LSHIFT and RSHIFT functionality and tests, and some refactoring has ¡been made in the testing classes, to make them more modular and flexible.
